File: thinking-categorically/monad-I/04-programming-definition /programming-definition.ts
// TypeScript: Generic types that add context
type Maybe<A> = A | null

function pureMaybe<A>(value: A): Maybe<A> {
    return value  // In Maybe, this just returns the value
}

// For Promise:
function purePromise<A>(value: A): Promise<A> {
    return Promise.resolve(value)
}

// TypeScript: flatMap/bind operation
function flatMap<A, B>(ma: Maybe<A>, f: (a: A) => Maybe<B>): Maybe<B> {
    if (ma === null) {
        return null           // short-circuit on failure
    }
    return f(ma)             // apply function to unwrapped value
}

// Helper function that might fail
function safeDivide(x: number, y: number): number | null {
    return y === 0 ? null : x / y;
}

// Chaining operations with flatMap
const computation =
    flatMap(pureMaybe(20), x =>
        flatMap(safeDivide(x, 4), y =>
            safeDivide(y, 2)));

// More readable with helper functions
function divideBy(divisor: number): (value: number) => number | null {
    return (value: number) => safeDivide(value, divisor);
}

const result =
    flatMap(flatMap(pureMaybe(20), divideBy(4)), divideBy(2));

// Result: 2.5

// Without monads, we'd need nested null checks:
// const step1 = safeDivide(20, 4);
// if (step1 !== null) {
//     const step2 = safeDivide(step1, 2);
//     if (step2 !== null) {
//         return step2;
//     }
// }
// return null;

File: thinking-categorically/monad-I/04-programming-definition /programming-definition.hs
module Main (main) where

-- Haskell: A type constructor that takes one type and produces another.
-- Maybe :: * -> *
-- Maybe Int wraps Int values that might be missing.
-- Maybe String wraps String values that might be missing.
--
-- Haskell: pure lifts values into the Maybe context:
-- pure 5       == Just 5
-- pure "hello" == Just "hello"

-- Problem: We have functions that return wrapped values
safeDivide :: Int -> Int -> Maybe Int
safeDivide _ 0 = Nothing
safeDivide x y = Just (x `div` y)

safeDivideBy :: Int -> Int -> Maybe Int
safeDivideBy divisor value = safeDivide value divisor

-- Chaining operations with bind
computation :: Maybe Int
computation =
    pure 20          -- Maybe Int
    >>= safeDivideBy 4 -- Maybe Int
    >>= safeDivideBy 2 -- Maybe Int
    -- Result: Just 2

-- Without monads, we'd need nested pattern matching:
-- case safeDivide 20 4 of
--   Nothing -> Nothing
--   Just x -> case safeDivide x 2 of
--     Nothing -> Nothing
--     Just y -> Just y

main :: IO ()
main = print computation
